apps/mongodb.go

```diff
@@ -0,0 +1,282 @@
+// limitations under the License.
+
+package apps
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+)
+
+type LoggingProcessorMongodb struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (*LoggingProcessorMongodb) Type() string {
+	return "mongodb"
+}
+
+func (p *LoggingProcessorMongodb) Components(tag, uid string) []fluentbit.Component {
+	c := []fluentbit.Component{}
+
+	c = append(c, p.JsonLogComponents(tag, uid)...)
+	c = append(c, p.RegexLogComponents(tag, uid)...)
+
+	return c
+}
+
+// JsonLogComponents are the fluentbit components for parsing log messages that are json formatted.
+// these are generally messages from mongo with versions greater than or equal to 4.4
+func (p *LoggingProcessorMongodb) JsonLogComponents(tag, uid string) []fluentbit.Component {
+	c := p.jsonParserWithTimeKey(tag, uid)
+
+	c = append(c, p.promoteWiredTiger(tag, uid)...)
+	c = append(c, p.severityParser(tag, uid)...)
+	c = append(c, p.renames(tag, uid)...)
+
+	return c
+}
+
+// jsonParserWithTimeKey requires promotion of the nested timekey for the json parser so we must
+// first promote the $date field from the "t" field before declaring the parser
+func (p *LoggingProcessorMongodb) jsonParserWithTimeKey(tag, uid string) []fluentbit.Component {
+	c := []fluentbit.Component{}
+
+	jsonParser := &confgenerator.LoggingProcessorJsonParser{
+		ParserShared: confgenerator.ParserShared{
+			TimeKey:    "time",
+			TimeFormat: "%Y-%m-%dT%H:%M:%S.%L%z",
+			Types: map[string]string{
+				"id":      "integer",
+				"message": "string",
+			},
+		},
+	}
+	jpComponents := jsonParser.Components(tag, uid)
+
+	parserComponent, filterComponent := jpComponents[0], jpComponents[1]
+
+	c = append(c, parserComponent, filterComponent)
+
+	tempPrefix := "temp_ts_"
+	timeKey := "time"
+	// have to bring $date to top level in order for it to be parsed as timeKey
+	// see https://github.com/fluent/fluent-bit/issues/1013
+	liftTs := fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":         "nest",
+			"Match":        tag,
+			"Operation":    "lift",
+			"Nested_under": "t",
+			"Add_prefix":   tempPrefix,
+		},
+	}
+
+	renameTsOption := fluentbit.NewHardRenameModifyOptions(fmt.Sprintf("%s$date", tempPrefix), timeKey, nil)
+	renameTs := renameTsOption.Component(tag)
+
+	c = append(c, liftTs, renameTs)
+
+	// IMPORTANT: now that we have lifted the json to top level
+	// we need to re-parse in order to properly set time at the
+	// parser level
+	c = append(c, filterComponent)
+
+	removeTimestamp := fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":   "modify",
+			"Match":  tag,
+			"Remove": timeKey,
+		},
+	}
+	c = append(c, removeTimestamp)
+
+	return c
+}
+
+func (p *LoggingProcessorMongodb) severityParser(tag, uid string) []fluentbit.Component {
```

Comment:
We tend to remove the `severity` field from `jsonPayload` after extracting the value and moving it to the top level `severity` field. Is that something we could do for MongoDB as well?